File: Top_100_Liked_Questions/169._Majority_Element/solution.py
class Solution:
    def majorityElementHashMap(self, nums: List[int]) -> int:
        # hashMap
        counts = collections.Counter(nums)
        return max(counts.keys(), key=counts.get)
    
    def majorityElement(self, nums: List[int]) -> int:
        # Boyer-Moore Voting Algorithm
        count = 0
        candidate = None
        
        for n in nums:
            if count == 0:
                candidate = n
            count += 1 if n == candidate else -1
        
        return candidate

File: Top_100_Liked_Questions/169._Majority_Element/solution.java
class Solution {
    public int majorityElement(int[] nums) {
        // voting algorithm
        int candidate = nums[0];
        int count = 1;

        for (int i=1; i<nums.length; i++) {
            if (count == 0)
                candidate = nums[i];

            count += nums[i] == candidate ? 1 : -1;
        }

        return candidate;
    }
}